dev/test/query.ts

```diff
@@ -3577,70 +3579,130 @@ describe('query resumption', () => {
-    // "startAt" of the given request.
-    function getStartAtDocumentIndex(
-      request: api.IRunQueryRequest
-    ): number | null {
-      const startAt = request.structuredQuery?.startAt;
-      const startAtValue = startAt?.values?.[0]?.referenceValue;
-      const startAtBefore = startAt?.before;
-      if (typeof startAtValue !== 'string') {
-        return null;
-      }
-      const docId = startAtValue.split('/').pop()!;
-      const docIdIndex = documentIds.indexOf(docId);
-      if (docIdIndex < 0) {
-        return null;
-      }
-      return startAtBefore ? docIdIndex : docIdIndex + 1;
+  // Return `numDocs` document responses, followed by an error response.
+  function* getDocResponsesFollowedByError(
+    documentIds: string[],
+    numDocs: number,
+    startAtEnd: boolean
+  ): Generator<api.IRunQueryResponse | Error> {
+    assert(numDocs <= documentIds.length);
+    const sliced = startAtEnd
+      ? documentIds.slice(-1 * numDocs)
+      : documentIds.slice(0, numDocs);
+    let runQueryResponses = sliced.map(documentId => result(documentId));
+    if (startAtEnd) {
+      runQueryResponses = runQueryResponses.reverse();
+    }
+    for (const runQueryResponse of runQueryResponses) {
+      yield runQueryResponse;
     }
+    const retryableError = new GoogleError('simulated retryable error');
+    retryableError.domain = RETRYABLE_ERROR_DOMAIN;
+    yield retryableError;
+  }
 
-    const RETRYABLE_ERROR_DOMAIN = 'RETRYABLE_ERROR_DOMAIN';
+  // Returns the documents from the given `documentIds` that match the `startAt`
+  // (or `endAt`) and the `limit` (if any) of the given request.
+  function* getDocResponsesForRequest(
+    request: api.IRunQueryRequest,
+    documentIds: string[]
+  ): Generator<api.IRunQueryResponse> {
+    if (request.structuredQuery?.startAt) {
+      const startAtDocumentIndex = getStartAtDocumentIndex(
+        request,
+        documentIds
+      );
+      if (startAtDocumentIndex === null) {
+        throw new Error('the request should specify a valid startAt');
+      }
 
-    // A mock replacement for Query._isPermanentRpcError which (a) resolves
-    // a promise once invoked and (b) treats a specific error "domain" as
-    // non-retryable.
-    function mockIsPermanentRpcError(err: GoogleError): boolean {
-      mockIsPermanentRpcError.invoked.resolve(true);
-      return err?.domain !== RETRYABLE_ERROR_DOMAIN;
-    }
-    mockIsPermanentRpcError.invoked = new Deferred();
+      let end: number | undefined = undefined;
+      if (request.structuredQuery?.limit?.value) {
+        end = startAtDocumentIndex + request.structuredQuery?.limit!.value;
+      }
 
-    // Return the first half of the documents, followed by a retryable error.
-    function* getRequest1Responses(): Generator<api.IRunQueryResponse | Error> {
       const runQueryResponses = documentIds
-        .slice(0, documentIds.length / 2)
+        .slice(startAtDocumentIndex, end)
```

Comment:
There is some code duplication at the end of block the "if" and "else" blocks. Consider factoring some of the common logic out, at least the loop that yields each value from a list and maybe too the mapping from documentId to response.